### ISSUE TITLE
Broadcast channel from RequestStorageAccess doesn't seem to work if not created before

### ISSUE THREAD

**Issue by ericalas-db** [privacysandbox/privacy-sandbox-dev-support#398]
Hello again,

Recently this ticket was updated, https://github.com/privacysandbox/privacy-sandbox-dev-support/issues/386, saying that the rSA BroadcastChannel API should work pre-3PCD 🎉 

In preparation, I have been working on this using the `Test Third Party Cookie Phaseout` flag enabled, until build 128 is released.

I am seeing an odd issue and am unsure if there is a misconfiguration on my end or a possible bug.

According to the [docs](https://developer.mozilla.org/en-US/docs/Web/API/Document/requestStorageAccess):
> requestStorageAccess() requests are automatically denied unless the embedded content is currently processing a user gesture such as a tap or click ([transient activation](https://developer.mozilla.org/en-US/docs/Glossary/Transient_activation)), or unless permission was already granted previously. If permission was not previously granted, they need to be run inside a user gesture-based event handler. 

Specifically
> unless permission was already granted previously

And it seems to be sort of working, based on my tests. If our application is opened in a tab **and** the Broadcast channel is already created, the iframe tab can be opened, and the channels can communicate. If, however, I open the iframe tab before the Broadcast channel is created, the channels cannot communicate. 

Steps that consistently work, assuming the user has already explicitly granted permissions  (example):
1. Open www.myapp.com, create a Broadcast channel `new BroadcastChannel("my-channel")`
2. Open www.embedding-myapp.com, this renders an iframe that loads www.myapp.com. myapp.com will automatically (i.e. on load) call rSA and use the Broadcast Channel, "my-channel"
3. the tab from step 1, fires a message
4. the iframe from tab2 successfully receives messages

To consistently have it not work:
1. Open www.myapp.com
2. Open www.embedding-myapp.com. myapp.com will automatically call rSA and use the Broadcast Channel, "my-channel"
3. go back to tab1 (www.myapp.com), and create a Broadcast channel
4. in tab from step 1, fire a message
4. the iframe from tab2 **does not** receive the message

Furthermore, if the iframe tab explicitly presses a button to call the rSA API, it works again

So tl;dr, can rSA Broadcast channels only work IF the app in a 1st party context has already created the channel?

Thank you for reading and your help!

**Comment by arichiv** [comment]
I'm unable to reproduce this using the demo at:
https://saa-beyond-cookies.glitch.me/ and https://saa-beyond-cookies-iframe.glitch.me/

**Comment by ericalas-db** [comment]
Interesting, okay, so it's not the order that they connect 🤔  I wonder if something is happening with the combination of using React, maybe?

Here is a hosting app that simply renders HTML that embeds using an iframe (with route `/embed`)
https://youthful-terrific-atrociraptor.glitch.me/

here is the react app:
https://mangrove-wood-gauge.glitch.me/
and the code for the `/embed` component: https://glitch.com/edit/#!/mangrove-wood-gauge?path=src%2FEmbedded.jsx%3A1%3A0

I have verified this on my machine and my colleague's machine. If you open `youthful-terrific` first and press "request permissions manually" accept the prompt

Next open `mangrove-wood` and press the "Send Message to iframe" button. `youthful-terrific` _should_ show a message with the timestamp. I have found that refreshing `youthful-terrific` and then pressing the "send message" button results in no messages being captured by the iframe. It's not an exact science, sometimes it works, but others it does not, for me it feels like a 50/50 chance of it working (also not scientific, but after refreshing waiting ~5s seems to be a trigger for it to not work)

here is a video showing how certain refresh combinations result in it not working and other times it does: https://streamable.com/s6b4wa

Few things to note:
in the `Embedded.jsx` file, inside a `useEffect` I am calling the rSA, which seems to "work" i.e. no permission errors and the broadcast channel can be accessed from the return of rSA

Thank you!

**Comment by arichiv** [comment]
I am seeing it start working, but then stop if I wait and try again after a few seconds. I know this doesn't happen in the demo I made, so I have to assume something about the way the BroadcastChannel is being stored in the react app is causing it to be cleared/garbage-collected. Can you try reproducing this using the OT disabling third-party storage partitioning (https://developer.chrome.com/origintrials/#/view_trial/568016503002103809) and just using broadcast channels without rSA? If this is still an issue then, it's something with react and not rSA

**Comment by ericalas-db** [comment]
> I am seeing it start working, but then stop if I wait and try again after a few seconds. I know this doesn't happen in the demo I made, so I have to assume something about the way the BroadcastChannel is being stored in the react app is causing it to be cleared/garbage-collected. Can you try reproducing this using the OT disabling third-party storage partitioning (https://developer.chrome.com/origintrials/#/view_trial/568016503002103809) and just using broadcast channels without rSA? If this is still an issue then, it's something with react and not rSA

Interesting! And for sure, I will run that test!

Just a few questions to make sure I do this correctly:
1. The OT states it's from Chrome 127-132; should I use the beta Chrome that is on 127? My standard Chrome is only 126
2. > just using broadcast channels without rSA

    Do you mean to keep using the demo app I sent and just call `new BroadcastChannel("test-channel")` instead of using rSA? If so, does this mean to imply that in the OT, iframes and top-level sites can just access the same storage, including BC, without needing to use rSA? (Thats what the description seems to imply, but just making sure :) )

In the meantime, I will assume that the OT can just access BC normally without rSA and begin my testing on my beta chrome!

Thank you for taking a look!

**Comment by arichiv** [comment]
1) https://developer.chrome.com/origintrials/#/view_trial/-8517432795264450559 works in M126
2) Yes

**Comment by ericalas-db** [comment]
> 1. https://developer.chrome.com/origintrials/#/view_trial/-8517432795264450559 works in M126
> 2. Yes

Awesome, thanks! just requested access (https://partnerissuetracker.corp.google.com/issues/354679100)

I assumed the web origin was the origin of the embedding app (https://mangrove-wood-gauge.glitch.me/)

once i get approval I'll start testing

**Comment by arichiv** [comment]
The request should be for the origin of the top-level domain that has an embed you want to unpartition. At least, that's the easiest way to do it.

**Comment by ericalas-db** [comment]
> The request should be for the origin of the top-level-frame that has sub-frame you want to unpartition. At least, that's the easiest way to do it.

Okay, I have done the following:
1. added the OT token in the HTML file that hosts the iframe (in this case, `youthful-terrific`)
2. updated `mangrove-wood` to create a BC using `new BroadcastChannel(test-channel)` - the function that creates and stores it is called on first load automatically without any user intervention. Note: I am still using the default react approach to store/use it
3. repeated my tests

I cannot reproduce the dropped messages from before, so _I don't believe_ React is doing something to the BC that is causing messages to be dropped 🤔 

You can see the altered version of the React app here: https://glitch.com/edit/#!/mangrove-wood-gauge?path=src%2FEmbedded.jsx%3A11%3A74

Note: I have removed the OT token from the HTML file, I wasn't sure if I should freely share this - but if you would like to test it, I can add it back and let you know.

**Comment by arichiv** [comment]
The OT token isn't secret, you can leave it in for testing

**Comment by arichiv** [comment]
One other thing you could try while using rSA, would be retaining the handle via react like you are with the channel. If that makes a difference it could inform the underlying issue.

**Comment by ericalas-db** [comment]
> One other thing you could try while using rSA, would be retaining the handle via react like you are with the channel. If that makes a difference it could inform the underlying issue.

I can certainly try that - I'll give it a go first thing in the morning

A follow-up question regarding https://github.com/privacysandbox/privacy-sandbox-dev-support/issues/386#issuecomment-2205847744 in Chrome 128: Does that mean that it would work as if I had the OT token in place?

**Comment by arichiv** [comment]
I was referring to the permission check in M128 for the rSA call, not anything about the underlying partitioning of a context (which is what the OT impacts).

**Comment by ericalas-db** [comment]
> One other thing you could try while using rSA, would be retaining the handle via react like you are with the channel. If that makes a difference it could inform the underlying issue.

Okay, this was the issue it appears. I have been testing this all morning, and it seems that retaining the handler fixes the problem. Thank you for your help and insight! 

I'll wait until the beta/release of 128 and retest with 3PCD off again! 

I believe that beta lands tomorrow, so i will test again once it does :)

**Comment by arichiv** [comment]
I'll follow up on this bug and try to resolve it for M129, retaining the handle should not be needed https://issues.chromium.org/u/1/issues/355018914

**Comment by ericalas-db** [comment]
> I'll follow up on this bug and try to resolve it for M129, retaining the handle should not be needed https://issues.chromium.org/u/1/issues/355018914

Thanks so much, I even saw you have a PR up, amazing, thank you!!

I will close this issue, I have also run tests on the latest Chrome beta (128) and can confirm it works as expected, thank you :)
